File: java_nio/EchoServer.java
import java.io.IOException;
import java.net.InetSocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.SelectionKey;
import java.nio.channels.Selector;
import java.nio.channels.ServerSocketChannel;
import java.nio.channels.SocketChannel;
import java.util.Iterator;
import java.util.Set;

public class EchoServer {
	private static final int PORT_NUMBER = 9000;
	private static final int BUFFER_SIZE = 1024;

	public static void main(String[] args) throws IOException {
		ServerSocketChannel server = ServerSocketChannel.open();
		server.socket().bind(new InetSocketAddress(PORT_NUMBER));
		server.socket().setReuseAddress(true);
		server.configureBlocking(false);

		Selector selector = Selector.open();
		server.register(selector, SelectionKey.OP_ACCEPT);

		while (true) {
			int channelCount = selector.select();
			if (channelCount <= 0)
				continue;
			Set<SelectionKey> keys = selector.selectedKeys();
			Iterator<SelectionKey> iterator = keys.iterator();
			while (iterator.hasNext()) {
				SelectionKey key = iterator.next();
				iterator.remove();

				if (key.isAcceptable()) {
					SocketChannel client = server.accept();
					client.configureBlocking(false);
					client.register(selector, SelectionKey.OP_READ, ByteBuffer.allocate(BUFFER_SIZE));
				} else if (key.isReadable()) {
					SocketChannel client = (SocketChannel) key.channel();
					ByteBuffer buffer = (ByteBuffer)key.attachment();
					if (client.read(buffer) < 0) {
						key.cancel();
						client.close();
					} else {
						buffer.flip(); // read from the buffer
						client.write(buffer);
						buffer.clear(); // write into the buffer            }
					}
				}
			}
		}
	}
}

File: kotlin_nio/echoserver.kt
import java.io.IOException
import java.net.InetSocketAddress
import java.nio.ByteBuffer
import java.nio.channels.SelectionKey
import java.nio.channels.Selector
import java.nio.channels.ServerSocketChannel
import java.nio.channels.SocketChannel

@Throws(IOException::class)
fun main(args: Array<String>) {
    val PORT_NUMBER = 9000
    val BUFFER_SIZE = 1024

    val selector = Selector.open()

    val channel = ServerSocketChannel.open()
    val address = InetSocketAddress("localhost", PORT_NUMBER)
    channel.apply {
        bind(address)
        configureBlocking(false)
        register(selector, channel.validOps())
    }

    while (true) {
        val channelCount = selector.select()
		if (channelCount <= 0)
			continue
        val keys = selector.selectedKeys()
        val iterator = keys.iterator()
        while (iterator.hasNext()) {
            val key = iterator.next()
            iterator.remove()

            if (key.isAcceptable) {
                channel.accept().apply {
                    configureBlocking(false)
                    register(selector, SelectionKey.OP_READ, ByteBuffer.allocate(BUFFER_SIZE))
                }
            } else if (key.isReadable) {
                val client = key.channel() as SocketChannel
                val buffer = key.attachment() as ByteBuffer
                if (client.read(buffer) < 0) {
                    key.cancel();
                    client.close();
                } else {
                    buffer.flip(); // read from the buffer
                    client.write(buffer);
                    buffer.clear(); // write into the buffer            }
                }
            }
        }
    }
}
